design/20230222_tinkerbell_crd_refactor.md

```diff
@@ -0,0 +1,520 @@
+	// ActionStateFailed state. Typically, this is an elaboration on the Reason. It is not
+	// relevant when the State field is not ActionStateFailed.
+	Message string `json:"message,omitempty"`
+}
+
+// State describes the point in time state of a Workflow.
+type WorkflowState string
+
+const (
+	WorkflowStatePending   WorkflowState = "Pending"
+	WorkflowStateRunning   WorkflowState = "Running"
+	WorkflowStateSucceeded WorkflowState = "Succeeded"
+	WorkflowStateFailed    WorkflowState = "Failed"
+)
+
+// ActionState describes a point in time state of an Action.
+type ActionState string
+
+const (
+	ActionStatePending      ActionState = "Pending"
+	ActionStateRunning      ActionState = "Running"
+	ActionStateSucceeded    ActionState = "Succeeded"
+	ActionStateFailed       ActionState = "Failed"
+)
+```
+
+### Workflow state machine
+
+![Workflow state machine](https://raw.githubusercontent.com/chrisdoherty4/tinkerbell-roadmap/feat/crd-refactor/design/images/tinkerbell_crd_refactor/workflow_state_machine.png)
+
+### Action state machine
+
+![Action state machine](https://raw.githubusercontent.com/chrisdoherty4/tinkerbell-roadmap/feat/crd-refactor/design/images/tinkerbell_crd_refactor/action_state_machine.png)
+
+### Resource validation
+
+Each CRD will leverage [CEL](https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/) for all validations supported by CEL. For validations such as ensuring MAC address uniqueness on `Hardware` resources a webhook will be used. The webhook will be served from the `tink-controller` component.
+
+### Data and functions available during template rendering
+
+Templates will have access to a subset of `Hardware` data when they are rendered. Injecting `Hardware` data was outlined in a [previous proposal](https://github.com/tinkerbell/proposals/tree/e24b19a628c6b1ecaafd566667155ca5d6fd6f70/proposals/0028). The existing implementation only injects disk that will, based on this proposal, be sourced from the `.Hardware.StorageDevices` list. This will continue to be available for new templates.
+
+All existing, custom template functions detailed in [`template_funcs.go`](https://github.com/tinkerbell/tink/blob/main/internal/workflow/template_funcs.go) will continue to be available during template rendering.
+
+### Hegel API Changes
+
+Hegel will undergo a reduction in the endpoints it serves because the data is no longer available on the `Hardware` resource. Minimally, Hegel will serve the following endpoints:
+
+* `/2009-04-04/meta-data/instance-id`
+* `/2009-04-04/user-data`
+
+This ensures compatability with cloud-init that explores the API via the instance ID endpoint.
+
+## Migrating from v1alpha1 to v1alpha2
+
+Given the relative immaturity of the Tinkerbell API we will not provide any migration tooling from `v1alpha1` to `v1alpha2`. Users will be required to manually convert `Hardware` and `Template` resources. `Workflow` resources are considered ephemeral and can generally be deleted.
```

Comment:
I think we should write some very rudimentary script or such to help with this process. I would recommend this for only for hardware and templates, not workflows.

Reply:
Sounds good. That should be simple enough.

Reply:
Should be resolved.